maven_iuvs/echelle.py

```diff
@@ -0,0 +1,1292 @@
+        except ValueError:
+            if ech_isdark(fidx):
+                if verbose:
+                    print(f"{fidx['name']} is dark, continuing")
+                    print()
+                continue # of course there will be no darks for a dark
+
+            continue 
+            
+    return lights_and_darks, lights_missing_darks
+
+
+def choose_dark(fidx, dark_options):
+    """
+    Choose which dark to use from a list of dark options. If only one is available, that will be used. 
+    If more, then a choice will occur.
+
+    Parameters
+    ----------
+    fidx : dictionary
+           file metadata for the light observation that could utilize the darks in dark_options.
+    dark_options : list
+                   file metadata for all files that could serve as a dark.
+
+
+    Returns
+    ----------
+    chosen_dark : dictionary
+                 file metadata of dark to use.
+    """
+    if len(dark_options) == 0:
+        return None
+    elif len(dark_options) == 1:
+        return dark_options[0]
+    else: 
+        return dark_options[0] # TO DO: Make this more intelligent
+
+
+def find_dark_options(input_light_idx, idx_list_to_search):
+    """
+    Looks for darks matching the observation described by input_light_idx.
+
+    Parameters
+    ----------
+    input_light_idx : dictionary
+                      a dictionary entry of metadata for some observation
+    idx_list_to_search : list of dictionaries
+                         where each dictionary is the metadata for dark files
+
+    Returns
+    ----------
+    dark_options : list of dictionaries
+                   where each dictionary is the metadata for dark files that 
+                   match input_light_idx
+    """
+    if not ech_islight(input_light_idx):
+        raise ValueError('Input file index corresponds to a dark observation, cannot find matching dark.')
+    
+    half_orbit = datetime.timedelta(hours=2)
+    dark_options = [didx for didx in idx_list_to_search 
+                    if (np.abs(didx['datetime'] - input_light_idx['datetime']) < half_orbit
+                        and didx['binning'] == input_light_idx['binning']
+                        # and didx['mcp_gain'] == input_light_idx['mcp_gain']
+                        and didx['int_time'] == input_light_idx['int_time']
+                        # and iuvs_orbno_from_fname(didx['name']) == iuvs_orbno_from_fname(input_light_idx['name'])
+                        and iuvs_segment_from_fname(didx['name']) == iuvs_segment_from_fname(input_light_idx['name'])
+                        and ech_isdark(didx))]
+    
+    return dark_options
+
+
+def ech_isdark(fidx):
+    """
+    Identifies whether an echelle file contains dark integrations by checking the gain.
+
+    Parameters
+    ----------
+    fidx : dictionary
+           a single dictionary entry of metadata for some observation
+    Returns
+    ----------
+    True or False
+    """
+
+    return 'dark' in fidx['name']
```

Comment:
need to add a TODO to impose a stricter test of "actually a dark"

Reply:
Added

Reply:
Fixed by d81b483549da109ff45b25607960cdaf97a7345b